File: frontend/src/api/schedules.ts
export type Operation = {
  id: string;
  name: string;
  event_date: string;
  location_name: string;
  status: string;
};

export type VersionSummary = {
  id: string;
  revision: number;
  status: string;
  origin: string;
  total_score: number;
  created_at: string;
};

export type Assignment = {
  id: string;
  volunteer_id: string;
  volunteer_name: string;
  shift_id: string;
  shift_name: string;
  starts_at: string;
  ends_at: string;
  response_status: string;
  reason_codes: string[];
  distance_km: number | null;
};

export type Shift = {
  id: string;
  name: string;
  role_name: string;
  starts_at: string;
  ends_at: string;
  capacity: number;
};

export type ToolCall = {
  tool: string;
  status: string;
  timestamp: string;
};

export type VersionDetail = VersionSummary & {
  operation_id: string;
  score_details: Record<string, number>;
  conflicts: string[];
  explanations: string[];
  assignments: Assignment[];
  shifts: Shift[];
  agent_session_id: string | null;
  agent_request_id: string | null;
  tool_calls: ToolCall[];
};

export type VersionDiff = {
  base_version_id: string | null;
  target_version_id: string;
  added: string[];
  removed: string[];
};

const apiUrl = import.meta.env.VITE_API_URL ?? "http://localhost:8000";

async function request(path: string, token: string, init: RequestInit = {}) {
  const response = await fetch(`${apiUrl}${path}`, {
    ...init,
    headers: { Authorization: `Bearer ${token}`, ...init.headers },
  });
  if (!response.ok) {
    const body = (await response.json().catch(() => ({}))) as { detail?: string };
    throw new Error(body.detail ?? "Action planning impossible");
  }
  return response;
}

export async function fetchOperations(organizationId: string, token: string) {
  return (await request(`/api/v1/organizations/${organizationId}/operations`, token)).json() as Promise<Operation[]>;
}

export async function fetchVersions(operationId: string, token: string) {
  return (await request(`/api/v1/operations/${operationId}/schedule-versions`, token)).json() as Promise<VersionSummary[]>;
}

export async function fetchVersion(operationId: string, versionId: string, token: string) {
  return (await request(`/api/v1/operations/${operationId}/schedule-versions/${versionId}`, token)).json() as Promise<VersionDetail>;
}

export async function generateVersion(operationId: string, instruction: string, token: string) {
  return (await request(`/api/v1/operations/${operationId}/schedule-versions`, token, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ instruction: instruction.trim() || null }),
  })).json() as Promise<VersionDetail>;
}

export async function moveVolunteer(
  versionId: string,
  volunteerId: string,
  fromShiftId: string,
  toShiftId: string,
  token: string,
) {
  return (await request(`/api/v1/schedule-versions/${versionId}/manual-moves`, token, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      volunteer_id: volunteerId,
      from_shift_id: fromShiftId,
      to_shift_id: toShiftId,
    }),
  })).json() as Promise<VersionDetail>;
}

export async function fetchDiff(versionId: string, token: string) {
  return (await request(`/api/v1/schedule-versions/${versionId}/diff`, token)).json() as Promise<VersionDiff>;
}

File: backend/app/schemas/schedules.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class OperationSummary(BaseModel):
    id: UUID
    name: str
    event_date: date
    location_name: str
    status: str


class GenerateScheduleRequest(BaseModel):
    instruction: str | None = Field(default=None, max_length=1000)
    base_version_id: UUID | None = None


class ManualMoveRequest(BaseModel):
    volunteer_id: UUID
    to_shift_id: UUID
    from_shift_id: UUID | None = None


class AssignmentView(BaseModel):
    id: UUID
    volunteer_id: UUID
    volunteer_name: str
    shift_id: UUID
    shift_name: str
    starts_at: datetime
    ends_at: datetime
    response_status: str
    reason_codes: list[str] = Field(default_factory=list)
    distance_km: float | None = None


class ShiftView(BaseModel):
    id: UUID
    name: str
    role_name: str
    starts_at: datetime
    ends_at: datetime
    capacity: int


class ScheduleVersionSummary(BaseModel):
    id: UUID
    revision: int
    status: str
    origin: str
    total_score: int
    created_at: datetime


class ToolCallView(BaseModel):
    tool: str
    status: str
    timestamp: datetime


class ScheduleVersionDetail(ScheduleVersionSummary):
    operation_id: UUID
    score_details: dict[str, int]
    conflicts: list[str]
    explanations: list[str]
    assignments: list[AssignmentView]
    shifts: list[ShiftView]
    agent_session_id: str | None = None
    agent_request_id: str | None = None
    tool_calls: list[ToolCallView] = Field(default_factory=list)


class ScheduleDiff(BaseModel):
    base_version_id: UUID | None
    target_version_id: UUID
    added: list[str]
    removed: list[str]

